Zend/zend_verify_type_inference.h

```diff
@@ -0,0 +1,204 @@
+				ZEND_VERIFY_TYPE_INFERENCE_ERROR("mask 0x%x missing array type %d", type_mask, Z_TYPE_P(val));
+				break;
+			}
+
+			/* Don't check all elements of large arrays. */
+			if (++num_checked > 16) {
+				break;
+			}
+		} ZEND_HASH_FOREACH_END();
+	}
+}
+
+/* Clang reports false positive unused warnings. */
+#ifdef __clang__
+__attribute__((unused))
+#endif
+static void zend_verify_inference_use(zend_execute_data *execute_data, const zend_op *opline)
+{
+	if (opline->op1_use_type
+	 && (opline->op1_type & (IS_TMP_VAR|IS_VAR|IS_CV))
+	 && opline->opcode != ZEND_ROPE_ADD
+	 && opline->opcode != ZEND_ROPE_END) {
+		zend_verify_type_inference(EX_VAR(opline->op1.var), opline->op1_use_type, opline->op1_type, execute_data, opline, "op1_use");
+	}
+	if (opline->op2_use_type
+	 && (opline->op2_type & (IS_TMP_VAR|IS_VAR|IS_CV))) {
+		zend_verify_type_inference(EX_VAR(opline->op2.var), opline->op2_use_type, opline->op2_type, execute_data, opline, "op2_use");
+	}
+	if (opline->result_use_type
+	 && (opline->result_type & (IS_TMP_VAR|IS_VAR|IS_CV))) {
+		zend_verify_type_inference(EX_VAR(opline->result.var), opline->result_use_type, opline->result_type, execute_data, opline, "result_use");
+	}
+}
+
+/* Clang reports false positive unused warnings. */
+#ifdef __clang__
+__attribute__((unused))
+#endif
+static void zend_verify_inference_def(zend_execute_data *execute_data, const zend_op *opline)
+{
+	if (EG(exception)) {
+		return;
+	}
+	if (opline->op1_def_type
+	 && (opline->op1_type & (IS_TMP_VAR|IS_VAR|IS_CV))
+	 // array is actually changed by the the following instruction(s)
+	 && opline->opcode != ZEND_FETCH_DIM_W
+	 && opline->opcode != ZEND_FETCH_DIM_RW
+	 && opline->opcode != ZEND_FETCH_DIM_FUNC_ARG
+	 && opline->opcode != ZEND_FETCH_LIST_W) {
+		zend_verify_type_inference(EX_VAR(opline->op1.var), opline->op1_def_type, opline->op1_type, execute_data, opline, "op1_def");
+	}
+	if (opline->op2_def_type
+	 && (opline->op2_type & (IS_TMP_VAR|IS_VAR|IS_CV))
+	 /* ZEND_FE_FETCH_R[W] does not define a result in the last iteration. */
+	 && opline->opcode != ZEND_FE_FETCH_R
+	 && opline->opcode != ZEND_FE_FETCH_RW) {
+		zend_verify_type_inference(EX_VAR(opline->op2.var), opline->op2_def_type, opline->op2_type, execute_data, opline, "op2_def");
+	}
+	if (opline->result_def_type
+	 && (opline->result_type & (IS_TMP_VAR|IS_VAR|IS_CV))
+	 && opline->opcode != ZEND_ROPE_INIT
+	 && opline->opcode != ZEND_ROPE_ADD
+	 /* Some jump opcode handlers don't set result when it's never read. */
+	 && opline->opcode != ZEND_JMP_SET
+	 && opline->opcode != ZEND_JMP_NULL
+	 && opline->opcode != ZEND_COALESCE
+	 && opline->opcode != ZEND_ASSERT_CHECK
+	 /* Smart branches may not declare result. */
+	 && !zend_is_smart_branch(opline)
+	 /* Calls only initialize result when returning from the called function. */
+	 && opline->opcode != ZEND_DO_FCALL
+	 && opline->opcode != ZEND_DO_ICALL
+	 && opline->opcode != ZEND_DO_UCALL
+	 && opline->opcode != ZEND_DO_FCALL_BY_NAME
```

Comment:
I didn't understand this comment.
Do you mean CALLs perform stackless control transfer to called functions?
``ICALL`` performs call and return. It should be a problem to check its result.

Reply:
This is referencing function calls without re-entry. They will trigger `VM_TRACE_OP_END` before the function was actually called, because we will only dispatch to the function later. Instead of verifying the type here, we will verify it when returning.

Reply:
What about ICALL?

Reply:
@dstogov Ah, sorry. Yes, that indeed should not be necessary. Thanks!